### PR TITLE
[structured config] Allow specifying unconfigured struct config resources

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

Allows using structured-config based resources and IO managers without binding config at defintiions-time, e.g.:

```python

class MyResource(Resource):
    a_string: str

@asset
def my_asset(res: MyResource):
    print(res.a_string)


defs = Definitions(
    assets=[my_asset],
    resources={"res": MyResource}
)
```
Screen Shot 2023-01-09 at 12 40 57 PM

## Test Plan

Add unit tests for resources, IO managers.

### CODE REVIEW

**Comment by benpankow** [comment]
Current dependencies on/for this PR:
* master
  * **PR #11589** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11589" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/11589?utm_source=stack-comment).

File: python_modules/dagster/dagster/_config/structured_config.py [hunk 1]
```diff
@@ -219,6 +258,36 @@ def create_io_manager_to_pass_to_user_code(self, context) -> IOManager:
...
+    ):
+        schema = infer_schema_from_config_class(io_manager)
+
+        def resource_fn(context: InitResourceContext):
+            instantiated = io_manager(**context.resource_config)
+
+            if is_io_manager_context_provided(instantiated):
+                return instantiated.resource_fn(context)
+            else:
+                return instantiated.resource_fn()
```
Comment (schrockn) [inline python_modules/dagster/dagster/_config/structured_config.py:282]: Seems like we can eliminate a bunch of code duplication here by making `is_io_manager_context_provided` and `is_context_provided`. Might have to do some shenanigans for make pyright happy but this feels like too much duplication between `UnconfiguredStructuredIOManager` and `UnconfiguredStructuredResource` to me. Even just a helpful function.

File: python_modules/dagster/dagster/_core/execution/build_resources.py [hunk 2]
```diff
@@ -124,6 +130,10 @@ def wrap_resources_for_execution(
     for resource_key, resource in resources.items():
         if isinstance(resource, ResourceDefinition):
             resource_defs[resource_key] = resource
+        elif is_structured_resource_cls(resource):
+            resource_defs[resource_key] = UnconfiguredStructuredResource(resource)
+        elif is_structured_io_manager_cls(resource):
+            resource_defs[resource_key] = UnconfiguredStructuredIOManager(resource)
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/execution/build_resources.py:136]: can we move this to a structured resource-only codepath?

**Review by schrockn — CHANGES_REQUESTED** [review]
See inline comments

File: python_modules/dagster/dagster/_core/execution/build_resources.py (hunk 2 above)
Comment (benpankow) [reply python_modules/dagster/dagster/_core/execution/build_resources.py:136]: Could you elaborate on what you mean? The `is_structured_X_cls` functions do restrict these statements to structured resources only. Did you mean Definitions-only codepath?

File: python_modules/dagster/dagster/_config/structured_config.py (hunk 1 above)
Comment (benpankow) [reply python_modules/dagster/dagster/_config/structured_config.py:282]: tidied up the implementations a bit, which lets us remove the context guards

**Pull request closed without merging**